src/plugins/controls/public/control_types/options_list/options_list_embeddable.tsx

```diff
@@ -111,38 +126,85 @@ export class OptionsListEmbeddable extends Embeddable<OptionsListEmbeddableInput
     );
 
     // push searchString changes into a debounced typeahead subject
-    this.typeaheadSubject = new Subject<string>();
     const typeaheadPipe = this.typeaheadSubject.pipe(
       tap((newSearchString) => (this.searchString = newSearchString)),
       debounceTime(100)
     );
 
     // fetch available options when input changes or when search string has changed
     this.subscriptions.add(
-      merge(dataFetchPipe, typeaheadPipe).subscribe(this.fetchAvailableOptions)
+      merge(dataFetchPipe, typeaheadPipe)
+        .pipe(skip(1)) // Skip the first input update because options list query will be run by initialize.
+        .subscribe(this.runOptionsListQuery)
     );
 
-    // build filters when selectedOptions change
+    // build filters when selectedOptions or invalidSelections change
     this.subscriptions.add(
-      this.getInput$()
+      this.componentStateSubject$
         .pipe(
-          debounceTime(400),
-          distinctUntilChanged((a, b) => isEqual(a.selectedOptions, b.selectedOptions)),
+          debounceTime(100),
+          distinctUntilChanged((a, b) => isEqual(a.validSelections, b.validSelections)),
           skip(1) // skip the first input update because initial filters will be built by initialize.
         )
         .subscribe(() => this.buildFilter())
     );
+
+    /**
+     * when input selectedOptions changes, check all selectedOptions against the latest value of invalidSelections.
+     **/
+    this.subscriptions.add(
+      this.getInput$()
+        .pipe(distinctUntilChanged((a, b) => isEqual(a.selectedOptions, b.selectedOptions)))
+        .subscribe(({ selectedOptions: newSelectedOptions }) => {
+          if (!newSelectedOptions || isEmpty(newSelectedOptions)) {
+            this.updateComponentState({
+              validSelections: [],
+              invalidSelections: [],
+            });
+            return;
+          }
+          const { invalidSelections } = this.componentStateSubject$.getValue();
+          const newValidSelections: string[] = [];
+          const newInvalidSelections: string[] = [];
+          for (const selectedOption of newSelectedOptions) {
+            if (invalidSelections?.includes(selectedOption)) {
+              newInvalidSelections.push(selectedOption);
+              continue;
```

Comment:
nit: Why not just use if/else vs continue?

Reply:
I guess continue was just there for style :sparkles:

I've used if else instead, because it reads a lot better. Good call!